torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,116 @@
+            missing_keys, unexpected_keys, error_msgs):
+        """load_state_dict pre-hook function for lazy buffers and parameters.
+
+        The purpose of this hook is to adjust the current state and/or
+        ``state_dict`` being loaded so that a module instance serialized in
+        both un/initialized state can be deserialized onto both un/initialized
+        module instance.
+        See comment in ``torch.nn.Module._register_load_state_dict_pre_hook``
+        for the details of the hook specification.
+        """
+        local_state = {k: v for k, v in self._parameters.items() if v is not None}
+        for name, param in local_state.items():
+            key = prefix + name
+            if key in state_dict:
+                input_param = state_dict[key]
+                if isinstance(param, UninitializedParameter): 
+                    # The current parameter is not initialized but the one being loaded one is
+                    # create a new parameter based on the uninitialized one
+                    if not isinstance(input_param, UninitializedParameter):
+                        with torch.no_grad():
+                            param.materialize(input_param.shape)
+
+    def initialize_parameters(self, *args, **kwargs):
+        r"""Initialize parameters according to the input batch properties.
+        This adds an interface to isolate parameter initialization from the
+        forward pass when doing parameter shape inference.
+        """
+        raise NotImplementedError('initialize_parameters is not implemented for {}'.format(self.__class__.__name__))
+
+    def has_uninitialized_params(self):
+        r"""Check if a module has parameters that are not initialized
+        """
+        # This is to avoid the JIT to track this parameter and force
+        # custom modules __setstate__ to add it
+        params = self._parameters.values()
+        for param in itertools.chain(params):
+            if isinstance(param, (UninitializedParameter)):
+                return True
+        return False
+
+    def _infer_parameters(self, module, input):
+        r"""Infers the size and initializes the parameters according to the
+        provided input batch.
+        Given a module that contains parameters that were declared inferrable
+        using :class:`torch.nn.parameter.ParameterMode.Infer`, runs a forward pass
+        in the complete module using the provided input to initialize all the parameters
+        as needed.
+        The module is set into evaluation mode before running the forward pass in order
+        to avoid saving statistics or calculating gradients
+        """
+        self.initialize_parameters(*input) 
+        if module.has_uninitialized_params():
+            raise RuntimeError('module {} has not been fully initialized'.format(self._get_name()))
+        self._initialize_hook.remove()
+        if self.cls_to_become is not None:
+            self.__class__ = self.cls_to_become
```

Comment:
Curious: would it work in general if the lazy module class is actually the class I want to use after the module has been initialized? (e.g. for a library where all modules are lazy and there is no "eager" module to replace with)

I might be overlooking some test cases, but it would be really awesome to highlight that this is possible :)

Reply:
if you dont override this parameter, it will remain as a lazy module😃

Reply:
Great idea! I just added it :}